zkevm-circuits/src/evm_circuit/execution/begin_tx.rs

```diff
@@ -137,15 +161,73 @@ impl<F: Field> ExecutionGadget<F> for BeginTxGadget<F> {
         );
 
         // Transfer value from caller to callee
+        // Use cb.query_word and BatchedIsZeroGadget as TransferWithGasFeeGadget
+        // expects words instead of expressions for tx_value and gas_fee
+        let effective_tx_value = cb.query_word();
+        let effective_gas_fee = cb.query_word();
+        let effective_tx_value_and_gas_fee_are_zero = BatchedIsZeroGadget::construct(
+            cb,
+            [effective_tx_value.expr(), effective_gas_fee.expr()],
+        );
+        cb.condition(is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "effective_tx_value == 0 && effective_gas_fee == 0",
+                effective_tx_value_and_gas_fee_are_zero.expr(),
+                true.expr(),
+            );
+        });
+
+        cb.condition(1.expr() - is_tx_invalid.expr(), |cb| {
+            cb.require_equal(
+                "effective_tx_value == tx_value",
+                effective_tx_value.expr(),
+                tx_value.expr(),
+            );
+            cb.require_equal(
+                "effective_gas_fee == mul_gas_fee_by_gas",
+                effective_gas_fee.expr(),
+                mul_gas_fee_by_gas.product().expr(),
+            );
+        });
+
         let transfer_with_gas_fee = TransferWithGasFeeGadget::construct(
             cb,
             tx_caller_address.expr(),
             tx_callee_address.expr(),
-            tx_value.clone(),
-            mul_gas_fee_by_gas.product().clone(),
+            effective_tx_value.clone(),
+            effective_gas_fee.clone(),
             &mut reversion_info,
         );
 
+        // Verify transfer
+        let sender_balance_prev = transfer_with_gas_fee.sender.balance_prev();
+
+        let add_tx_value_and_mul_gas_fee_by_gas = cb.query_word();
+
+        // tx_value.int_value + gas_fee.int_value
+        let total_eth_cost = AddWordsGadget::construct(
+            cb,
+            [tx_value.clone(), mul_gas_fee_by_gas.product().clone()],
+            add_tx_value_and_mul_gas_fee_by_gas.clone(),
+        );
+
+        // balance_not_enough = sender_balance_prev.int_value < tx_value.int_value +
+        // gas_fee.int_value
+        let balance_not_enough = LtWordGadget::construct(
+            cb,
+            &sender_balance_prev,
+            &add_tx_value_and_mul_gas_fee_by_gas,
+        );
+
+        // invalid_tx = 1 - (1 - balance_not_enough) * (is_nonce_valid)
+        // prover should not give incorrect is_tx_invalid flag.
+        // instruction.constrain_equal(is_tx_invalid, invalid_tx)
+        cb.require_equal(
+            "is_tx_invalid is correct",
+            or::expr([balance_not_enough.expr(), not::expr(is_nonce_valid.expr())]),
+            is_tx_invalid.expr(),
+        );
```

Comment:
Gas not enough case still missing? https://github.com/smtmfft/zkevm-specs/blob/4cf8c8aa96453830c5440f1d6afac78997fbe8c6/src/zkevm_specs/evm/execution/begin_tx.py#L79